modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/SentStateChannelBinaryTrackingService.scala

```diff
@@ -0,0 +1,64 @@
+package org.tessellation.currency.l0.snapshot.services
+
+import cats.effect.{Async, Ref}
+import cats.syntax.all._
+
+import scala.collection.immutable.ListMap
+
+import org.tessellation.currency.l0.node.IdentifierStorage
+import org.tessellation.schema.GlobalIncrementalSnapshot
+import org.tessellation.security.signature.Signed
+import org.tessellation.statechannel.StateChannelSnapshotBinary
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.{NonNegInt, PosInt}
+
+trait SentStateChannelBinaryTrackingService[F[_]] {
+  def setPending(binary: Signed[StateChannelSnapshotBinary]): F[Unit]
+  def getPending: F[ListMap[Signed[StateChannelSnapshotBinary], NonNegInt]]
+  def getRetriable: F[Option[Signed[StateChannelSnapshotBinary]]]
+  def updateByGlobalSnapshot(globalSnapshot: GlobalIncrementalSnapshot): F[Unit]
+}
+
+object SentStateChannelBinaryTrackingService {
+  private val retryOrdinalDelay: PosInt = 3
+
+  def make[F[_]: Async](
+    identifierStorage: IdentifierStorage[F]
+  ): F[SentStateChannelBinaryTrackingService[F]] =
+    Ref
+      .of[F, ListMap[Signed[StateChannelSnapshotBinary], NonNegInt]](ListMap.empty[Signed[StateChannelSnapshotBinary], NonNegInt])
+      .map(make[F](_, identifierStorage))
+
+  private def make[F[_]: Async](
+    pendingR: Ref[F, ListMap[Signed[StateChannelSnapshotBinary], NonNegInt]],
+    identifierStorage: IdentifierStorage[F]
+  ): SentStateChannelBinaryTrackingService[F] =
+    new SentStateChannelBinaryTrackingService[F] {
+
+      private def incrementChecks(
+        pending: ListMap[Signed[StateChannelSnapshotBinary], NonNegInt]
+      ): ListMap[Signed[StateChannelSnapshotBinary], NonNegInt] =
+        ListMap(pending.view.mapValues(checks => NonNegInt.unsafeFrom(checks + 1)).toSeq: _*)
+
+      private def updateByGlobalSnapshotBinaries(binaries: List[Signed[StateChannelSnapshotBinary]]): F[Unit] =
+        pendingR.update { currPending =>
+          incrementChecks(currPending -- binaries.toSet)
+        }
+
+      def setPending(binary: Signed[StateChannelSnapshotBinary]): F[Unit] =
+        pendingR.update(_.updated(binary, NonNegInt.MinValue))
+
+      def getPending: F[ListMap[Signed[StateChannelSnapshotBinary], NonNegInt]] = pendingR.get
+
+      def getRetriable: F[Option[Signed[StateChannelSnapshotBinary]]] = getPending.map { pending =>
+        pending.find { case (_, checks) => checks >= retryOrdinalDelay }.map(_._1)
+      }
+
+      def updateByGlobalSnapshot(globalSnapshot: GlobalIncrementalSnapshot): F[Unit] =
```

Comment:
I would account for one race condition that I think may happen, which is: a node finishing a consensus with some delay in regards to other nodes, then setting the pending binary and at the same time processing GL0 snapshot containing the binary (when the pending binary is not yet set). Other non-delayed nodes may propose the binary to GL0 which may add it to the GL0 snapshot. And this global snapshot may be processed before the binary is locally set as pending. Effectively that binary will never be detected as accepted by GL0. What I think we need to do is discard (assume as accepted) all binaries that are below some binary detected as accepted. If binary with hash ABC is accepted on GL0 then every snapshot leading up to ABC had to be accepted as well. I would add such logic when updating accepted binaries e.g. find the highest index of binary found in GL0 snapshot and then take the slice above such index to remain as pending.